Move memory allocation out of elkan_L2_sse

bjzhjing opened the pull request:
elkan_L2_sse() does malloc/free each time when it's called by the loop in train_encoded() during data train for most of IVF index building. Memory allocation triggers the process of vma allocation, physical page allocation, page mapping setup and memory cgroup statistic update in kernel side, so more frequent allocation means more overhead. Move it out of the loop and do it only once when IndexFlatElkan is initialized to save CPU cycles.

issue #279

liliu-z commented:
/kind enhancement

alexanderguzhva commented on thirdparty/faiss/faiss/utils/distances.h:
```diff
@@ -422,6 +422,7 @@ void compute_PQ_dis_tables_dsub2(
  * @param y    database vectors, size ny * d
  * @param ids  result array ids
  * @param val  result array value
+ * @param data the pointer of memory for symmetric matrix data
```
Please call this variable `tmp_buffer` in order to stress out its meaning. It is not a data, it is a temporary storage.

bjzhjing replied:
Thanks for helping review, Alex! The change is done.

alexanderguzhva commented on thirdparty/faiss/faiss/IndexFlatElkan.cpp:
```diff
@@ -15,12 +15,15 @@
 namespace faiss {
 
-IndexFlatElkan::IndexFlatElkan(idx_t d, MetricType metric, bool is_cosine, bool use_elkan)
+IndexFlatElkan::IndexFlatElkan(idx_t d, MetricType metric, bool is_cosine, bool use_elkan, float* data)
```
?! `data` which is passed as a parameters is not used anywhere

bjzhjing replied:
Yeah, it's not required to be set by callers. Remove it from the interface.

alexanderguzhva commented on thirdparty/faiss/faiss/IndexFlatElkan.h:
```diff
@@ -29,9 +29,17 @@ namespace faiss {
 // Elkan algo was introduced into Knowhere in #2178, #2180 and #2258. 
 struct IndexFlatElkan : IndexFlat {
     bool use_elkan = true;
+    float* data = nullptr;
```
`std::vector<float> tmp_buffer_for_elkan;`

bjzhjing replied:
Changed it.

alexanderguzhva commented on thirdparty/faiss/faiss/IndexFlatElkan.h:
```diff
@@ -29,9 +29,17 @@ namespace faiss {
 
     explicit IndexFlatElkan(idx_t d, MetricType metric = METRIC_L2,
-                       bool is_cosine = false, bool use_elkan = true);
+                       bool is_cosine = false, bool use_elkan = true, float* data = nullptr);
```
I see no point in passing `float* data` to constructor. By design, it is a temporary buffer

bjzhjing replied:
Yes, remove it.

alexanderguzhva commented on thirdparty/faiss/faiss/utils/distances.cpp:
```diff
@@ -1155,13 +1155,18 @@ void elkan_L2_sse(
 
     const size_t bs_y = 1024;
-    float* data = (float*)malloc((bs_y * (bs_y - 1) / 2) * sizeof(float));
+    bool allocate = false;
```
Please use `std::unique_ptr<float[]>` + `new` for allocating a local temporary buffer

bjzhjing replied:
Call resize() to allocate the memory in constructor. Remove the checking here for use_elkan can help ensure the memory is allocated when come here.

bjzhjing commented:
@alexanderguzhva Hi Alex, I use smart pointer instead of std::vector to implement the tmp_buffer finally, for I IndexFlatElkan::search is defined with 'const', which means it's not allowed to modify the class variables. Thus, it's required to use const_cast<std::vector<float>&> for convert while passing tmp_buffer to elkan_L2_sse, there is address change, which make things a bit complex. Smart pointer can help avoid the issue and make the code more clean and safe.

Please help review. Thanks very much!

alexanderguzhva commented:
@bjzhjing Two more notes
1. Please add a comment about the situation to the top of `IndexFlatElkan` . Basically, in the provided form `IndexFlatElkan` is no longer thread-safe for the search. So, two threads cannot use 'IndexFlatElkan::search()' simultaneosly on the same instance. This is fine, bcz `IndexFlatElkan` is used only during the index construction. But the note about the implementation detail and why this was done should be included.
2. The needed size of the temporary buffer depends on the `const size_t bs_y = 1024;` magic constant inside  `elkan_L2_sse()`. So, notes on the sizes need to be added to the description of the `elkan_L2_sse()` implementation. Maybe, `bs_y` size needs to be exposed as a function parameter as well.

bjzhjing commented:
> @bjzhjing Two more notes
> 
> 1. Please add a comment about the situation to the top of `IndexFlatElkan` . Basically, in the provided form `IndexFlatElkan` is no longer thread-safe for the search. So, two threads cannot use 'IndexFlatElkan::search()' simultaneosly on the same instance. This is fine, bcz `IndexFlatElkan` is used only during the index construction. But the note about the implementation detail and why this was done should be included.
> 2. The needed size of the temporary buffer depends on the `const size_t bs_y = 1024;` magic constant inside  `elkan_L2_sse()`. So, notes on the sizes need to be added to the description of the `elkan_L2_sse()` implementation. Maybe, `bs_y` size needs to be exposed as a function parameter as well.

Greetings @alexanderguzhva! I'm just back from a bad cold, and sorry for the late response. I've addressed your above comments. Please help review, thanks!

PwzXxm commented:
/lgtm
/approve
